Make PR Jenkins jobs upload build artifacts to GCS

ixdy opened the pull request:
We already do this for GKE jobs, so it seems like it won't be a huge deal to do it for all jobs.

I mostly want to do this so we can get the `gcs-stage/$version` directory created (x-ref https://github.com/kubernetes/release/pull/166). 

There isn't an easy way to tell the release scripts to stage the files but not actually upload to GCS, but if we feel strongly about that, I guess I can go shave that yak, too.

cc @david-mcmahon 

x-ref https://github.com/kubernetes/test-infra/pull/885

ixdy commented:
I am a little worried about having every job upload to the same path. There are currently two GKE jobs uploading to the same path without issues, though, so maybe it won't be a problem?

spxtr commented:
I am also worried about that. Those two GKE jobs aren't exactly paragons of stability, and I can't help but wonder if some of their flakiness is due to this.

spxtr commented:
It might be a good time to look into making the infra required to build/upload as one job and then trigger test jobs.

fejta commented:
/lgtm

Might be worth talking about @spxtr and @rmmh to see if we can get pipelining into the PR testing roadmap -- aka we build once and then kick off jobs to do tests.

fejta commented:
FWIW I suspect that this will manifest itself in early termination of gcs downloads (if you upload a file while I'm downloading it then my download terminates unexpectedly), however gcs should recover from this.

The other thing I am concerned about is that does the path include a hash of the contents? It could be the fact that agents A and B have different dependencies (of Go, etc) that result in building slightly different binaries at the same git commit.

spxtr commented:
You will get slightly different binaries for the same git commit because we bake the build time into the binaries.

ixdy commented:
Ooh, I had forgotten about the build time embedded in the binaries. curse you, past-@ixdy.

I see basically three options:
 1. merge as-is, possibly making PR GKE builds more unstable
 2. upload artifacts for each build into a `$JOB_NAME` sub-directory on GCS
 3. refactor release scripts to separate local staging from GCS uploading

I'm inclined to go with option 2 as optimizes effort and safety.

rmmh commented:
#2 SGTM. Do we need to do anything to clean up old builds?

ixdy commented:
Updated, PTAL. I also updated the GKE jobs so that they rsync to non-conflicting paths in kubernetes-release-dev.

@rmmh we should probably set a TTL on gs://kubernetes-release-pull if we haven't already.

david-mcmahon commented:
Could the use of includes (sourced) here help with reducing dup across jobs?

ixdy commented:
@rmmh 
```console
$ gsutil lifecycle get gs://kubernetes-jenkins
{"rule": [{"action": {"type": "Delete"}, "condition": {"age": 90}}]}
$ gsutil lifecycle get gs://kubernetes-release-pull
gs://kubernetes-release-pull/ has no lifecycle configuration.
```

Anyone object to a 90d TTL for `gs://kubernetes-release-pull`? Or even 14d?

ixdy commented:
```console
gsutil lifecycle get gs://kubernetes-release-dev
{"rule": [{"action": {"type": "Delete"}, "condition": {"age": 60, "isLive": false}}]}
```

spxtr commented:
Oops. Lets set that to 14 days.

ixdy commented:
I'm not sure what's up with the lifecycle settings on kubernetes-release-dev, but there are lots of artifacts older than 60d there. I'm guessing the `isLive` setting is preventing stuff from being deleted.

ixdy commented:
Any thoughts on this PR? Still LG?

rmmh commented:
isLive:false means that it will *only* delete old versions of each object (kubernetes-release-dev is versioned). That's _probably_ not intended.

ixdy commented:
```console
$ gsutil lifecycle get gs://kubernetes-release-pull/
{"rule": [{"action": {"type": "Delete"}, "condition": {"age": 14}}]}
```

We should follow up on `gs://kubernetes-release-dev` somewhere else, probably. :)

@david-mcmahon we should probably discuss your suggestion in a different place, too.

fejta commented:
LGTM
